sycl/include/sycl/kernel_bundle.hpp

```diff
@@ -308,18 +308,19 @@ class kernel_bundle : public detail::kernel_bundle_plain,
   template <auto &SpecName>
   typename std::remove_reference_t<decltype(SpecName)>::value_type
   get_specialization_constant() const {
-    const char *SpecSymName = detail::get_spec_constant_symbolic_ID<SpecName>();
-    if (!is_specialization_constant_set(SpecSymName))
-      return SpecName.getDefaultValue();
-
     using SCType =
         typename std::remove_reference_t<decltype(SpecName)>::value_type;
 
-    std::array<char *, sizeof(SCType)> RetValue;
+    const char *SpecSymName = detail::get_spec_constant_symbolic_ID<SpecName>();
+    SCType Res{SpecName.getDefaultValue()};
+    if (!is_specialization_constant_set(SpecSymName))
+      return Res;
 
+    std::array<char, sizeof(SCType)> RetValue;
```

Comment:
Note for reviewers: This changes the type in the array. This is not directly related to this bug, but it avoids overallocation as we only need `sizeof(SCType)` bytes, while a `char *` is likely larger than a byte.

Reply:
This was reverted as it caused issues. This should be investigated separately.